File: ruby/day1/part1.rb
# frozen_string_literal: true

def parse_input(input)
  left, right = input.split("\n")
                     .map { |line| line.split.map(&:to_i) }
                     .transpose
  [left.sort, right.sort]
end

def part1(input)
  left, right = parse_input(input)
  left.zip(right)
      .sum { |l, r| (l - r).abs }
end

if __FILE__ == $PROGRAM_NAME
  input = File.read('input.txt')
  p part1(input)
end

File: elixir/day1/lib/part1.ex
defmodule Part1 do
  def parse_line(line) do
    line
    |> String.split(~r/\s+/, trim: true)
    |> Enum.map(fn [left_str, right_str] ->
      {String.to_integer(left_str), String.to_integer(right_str)}
    end)
  end

  def parse_input(input) do
    {lefts, rights} =
      input
      |> String.split("\n", trim: true)
      |> Enum.map(&parse_line/1)
      |> Enum.unzip()

    {Enum.sort(lefts), Enum.sort(rights)}
  end

  def part1(input) do
    {left_list, right_list} = parse_input(input)

    Enum.zip(left_list, right_list)
    |> Enum.map(fn {l, r} -> abs(l - r) end)
    |> Enum.sum()
  end
end

File.read!(~c"input.txt")
|> Part1.part1()
|> IO.inspect(label: "part1")
